### PR TITLE
Vb 2657 admin api endpoint to migrate visits between session templates

### PR DESCRIPTION
**Pull request by dhirajshettymoj**
## What does this pull request do?
Admin api endpoint to migrate visits between session templates

## What is the intent behind these changes?

Admin API improvements

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/service/SessionTemplateService.kt [hunk 1]
```diff
@@ -482,6 +501,21 @@ class SessionTemplateService(
...
   }
+
+  fun moveSessionTemplateVisits(fromSessionTemplateReference: String, toSessionTemplateReference: String, fromDate: LocalDate): Int {
+    val fromSessionTemplate = SessionTemplateDto(getSessionTemplate(fromSessionTemplateReference))
+    val toSessionTemplate = SessionTemplateDto(getSessionTemplate(toSessionTemplateReference))
+    val fromSessionTemplateVisitStats = getSessionTemplateVisitStats(fromSessionTemplate.reference, RequestSessionTemplateVisitStatsDto(fromDate))
+    val toSessionTemplateVisitStats = getSessionTemplateVisitStats(toSessionTemplate.reference, RequestSessionTemplateVisitStatsDto(fromDate))
+
+    visitMoveValidator.validateMoveSessionTemplateVisits(fromSessionTemplate, fromSessionTemplateVisitStats, toSessionTemplate, toSessionTemplateVisitStats, fromDate)
+
```
Comment (AledWEvans) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/service/SessionTemplateService.kt:512]: Would move :
```
    val fromSessionTemplateVisitStats = getSessionTemplateVisitStats(fromSessionTemplate.reference, RequestSessionTemplateVisitStatsDto(fromDate))
    val toSessionTemplateVisitStats = getSessionTemplateVisitStats(toSessionTemplate.reference, RequestSessionTemplateVisitStatsDto(fromDate))
```
into `validateMoveSessionTemplateVisits`

so we encapsulate it correctly

File: src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/utils/SessionDatesUtil.kt [hunk 2]
```diff
@@ -41,19 +42,20 @@ class SessionDatesUtil {
...
   }
 
   private fun getValidFromMonday(validFromDate: LocalDate): LocalDate {
     // This has been added just encase someone wants the session template start date other than the start of week.
     // Therefore, this use of validFromMonday will allow the bi-weekly to still work.
-    if (validFromDate.dayOfWeek != MONDAY) {
-      return validFromDate.with(TemporalAdjusters.previous(MONDAY))
-    }
-    return validFromDate
+    return validFromDate.with(TemporalAdjusters.previousOrSame(MONDAY))
```
Comment (AledWEvans) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/utils/SessionDatesUtil.kt:58]: nice little refactor... previousOrSame

File: src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/utils/SessionTemplateComparator.kt [hunk 3]
```diff
@@ -120,42 +103,27 @@ class SessionTemplateComparator(private val sessionDatesUtil: SessionDatesUtil)
...
-
-      hasLocationMatch(newSessionLocations, existingSessionLocations)
-    } else {
-      false
-    }
+    val existingSessionLocations = sessionTemplateUtil.getPermittedSessionLocations(existingSessionDetails.permittedLocationGroups)
+    val newSessionLocations = sessionTemplateUtil.getPermittedSessionLocations(newSessionDetails.permittedLocationGroups)
+
+    return sessionLocationMatcher.hasAnyMatch(newSessionLocations, existingSessionLocations) ||
+      sessionLocationMatcher.hasAnyMatch(existingSessionLocations, newSessionLocations)
```
Comment (AledWEvans) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/utils/SessionTemplateComparator.kt:110]: be careful changing code out side of ticket scope,...

File: src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/utils/SessionTemplateUtil.kt [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+import uk.gov.justice.digital.hmpps.visitscheduler.dto.sessions.location.SessionLocationGroupDto
+import uk.gov.justice.digital.hmpps.visitscheduler.model.entity.session.category.PrisonerCategoryType
+import uk.gov.justice.digital.hmpps.visitscheduler.model.entity.session.incentive.IncentiveLevel
+import java.util.*
+import java.util.stream.Collectors
+
+@Component
+class SessionTemplateUtil {
+  val sessionCapacityComparator: Comparator<SessionCapacityDto> =
+    Comparator { o1: SessionCapacityDto, o2: SessionCapacityDto ->
```
Comment (AledWEvans) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/utils/SessionTemplateUtil.kt:18]: What is o1 and o2, please use meaning full names, like 
sessionCapacityOne and sessionCapacityTwo

File: src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/utils/validators/SessionTemplateVisitMoveValidator.kt [hunk 5]
```diff
@@ -0,0 +1,212 @@
...
+    allowedToSessionCapacity: SessionCapacityDto,
+  ): List<LocalDate> {
+    val exceededSessionCapacityDates = mutableListOf<LocalDate>()
+    val toSessionVisits = toSessionTemplateVisitStats.visitsByDate?.associateBy { it.visitDate }
+    if (toSessionVisits != null) {
+      fromSessionTemplateVisitStats.visitsByDate?.forEach {
+        val toSessionTemplateVisitsForDate = toSessionVisits[it.visitDate]
+        if (toSessionTemplateVisitsForDate != null) {
+          val totalCapacityForVisitDate = sessionTemplateUtil.getTotalCapacity(it.visitCounts, toSessionTemplateVisitsForDate.visitCounts)
+          if (sessionTemplateUtil.sessionCapacityComparator.compare(allowedToSessionCapacity, totalCapacityForVisitDate) < 0) {
```
Comment (AledWEvans) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/utils/validators/SessionTemplateVisitMoveValidator.kt:183]:
```
val totalCapacityForVisitDate = sessionTemplateUtil.getTotalCapacity(it.visitCounts, toSessionTemplateVisitsForDate.visitCounts)
if (sessionTemplateUtil.sessionCapacityComparator.compare(allowedToSessionCapacity, totalCapacityForVisitDate) < 0) {
```

could this be put in the sessionTemplateUtil : 
```
if (sessionTemplateUtil.isCapacityExceeded(it.visitCounts, 
     toSessionTemplateVisitsForDate.visitCounts,
     allowedToSessionCapacity)) {
```
Something like that

**Review by AledWEvans — CHANGES_REQUESTED** [review]
Very good work, some comments that may need action

File: src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/service/SessionTemplateService.kt (hunk 1 above)
Comment (dhirajshettymoj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/service/SessionTemplateService.kt:510]: moved to a new method in SessionTemplateService.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/utils/SessionTemplateComparator.kt (hunk 3 above)
Comment (dhirajshettymoj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/utils/SessionTemplateComparator.kt:110]: mandated by the code changes to this PR

File: src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/utils/SessionTemplateUtil.kt (hunk 4 above)
Comment (dhirajshettymoj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/utils/SessionTemplateUtil.kt]: Updated

File: src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/utils/validators/SessionTemplateVisitMoveValidator.kt (hunk 5 above)
Comment (dhirajshettymoj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/utils/validators/SessionTemplateVisitMoveValidator.kt]: Updated code to use + operator and added isCapacityExceeded to SessionTemplateUtil as suggested.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/repository/VisitRepository.kt [hunk 6]
```diff
@@ -195,7 +195,7 @@ interface VisitRepository : JpaRepository<Visit, Long>, JpaSpecificationExecutor
   @Modifying
   @Query(
     "Update visit set session_template_reference = :newSessionTemplateReference " +
-      "WHERE (visit_status = 'BOOKED' OR visit_status = 'RESERVED')  AND " +
+      "WHERE (visit_status = 'BOOKED' OR visit_status = 'RESERVED' OR visit_status = 'CHANGING')  AND " +
```
Comment (AledWEvans) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/repository/VisitRepository.kt:198]: could use visit_status in (' your call?

**Review by AledWEvans — APPROVED** [review]
LGTM one minor comment, your call

File: src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/repository/VisitRepository.kt (hunk 6 above)
Comment (dhirajshettymoj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/repository/VisitRepository.kt]: updated

**Review by AledWEvans — APPROVED** [review]
LGTM

**Pull request merged**